src/typechecker.ts

```diff
@@ -77,17 +81,31 @@ const getExpressionVisitor =
         const argType = typeFromExpr(arg, spec.biblio);
         const paramType = typeFromExprType(param.type);
 
-        // often we can't infer the argument precisely, so we check only that the intersection is nonempty rather than that the argument type is a subtype of the parameter type
-        const intersection = meet(argType, paramType);
-
-        if (
-          intersection.kind === 'never' ||
-          (intersection.kind === 'list' &&
-            intersection.of.kind === 'never' &&
-            // if the meet is list<never>, and we're passing a concrete list, it had better be empty
-            argType.kind === 'list' &&
-            argType.of.kind !== 'never')
-        ) {
+        const error = getErrorForUsingTypeXAsTypeY(argType, paramType);
+        if (error != null) {
+          let hint: string;
+          switch (error) {
+            case 'number-to-math': {
+              hint =
+                '\nhint: you passed an ES language Number, but this position takes a mathematical value';
+              break;
+            }
+            case 'math-to-number': {
+              hint =
+                '\nhint: you passed a mathematical value, but this position takes an ES language Number';
+              break;
+            }
+            case 'math-to-bigint': {
+              hint =
+                '\nhint: you passed a mathematical value, but this position takes an ES language BigInt';
+              break;
+            }
+            case 'other': {
```

Comment:
Instead of testing for `null` above, either use a `default` or (if you're looking to get an exhaustivity check from TS), fallthrough with `case null:` here.

Reply:
The test for `null` guards the whole error-producing logic, not just this switch case.